[RDY] Luadoc: Lua doxygen docstrings

KillTheMule opened the pull request:
First shot at generating lua API docs via doxygen. You can run `doxygen src/Doxyfile`, and it will pick up the lua files, and also the doc comments I added in `src/nvim/lua/vim.lua`.

Questions:
- I probably need to modify `gen_api_doc.py`, but I have no idea what's going on with the `INPUT_FILTER` argument to doxygen. Since we need a different filter for the lua files, we probably need to use `FILTER_PATTERNS` instead (can't use both). What's going on here, what do I need to do?
- I don't really understand `lua2dox.lua`, but I can hack it a bit. In particular, the chang to `macroStr` was simply to let it run without error, not sure if it's even remotely correct. Does anyone understand doxygen enough to tell me?
- Generally, how should those doc comments look like? I simply used what works right now, but I assume we have some leeway. Should we try to imitate the style from the C files?
- What do we need to do integration wise? In particular, I don't think this runs on windows, we need doxygen to pick up the bat file. I assume we can do it from `gen_api_doc.py`, but is there something else?

justinmk commented on src/nvim/lua/vim.lua:
```diff
@@ -9,6 +9,9 @@ end
 -- Used by nvim_get_proc() as a fallback.
+--!brief Gets the stuff for pid
+--!param pid The pid
+--!return A table
```
any idea if there are other syntaxes to choose from, besides `--!` ? (not too important, but I know doxygen has a bunch, would be nice if `--@` worked since that's similar to what we use in the C files.)

KillTheMule replied:
Yes, very much, that's what I meant by "how should those doc comments look like". I'll make it `--@return` then.

justinmk commented:
- I'd say don't bother with `gen_api_doc.py` for now. Can revisit it later. 
- What output does this produce currently?
- If `lua2dox.lua` is "generalized" we could include it in `runtime/lua/` . What do you think? 

> What do we need to do integration wise? In particular, I don't think this runs on windows, we need doxygen to pick up the bat file. 

We don't need it to run on Windows, because this is not part of the build. We generate docs manually, and may later add it as a CI task. I would delete the `.bat` file.

KillTheMule commented:
> What output does this produce currently?

The lua docs are mashed into the C docs, because the way this works is by creating a "dummy" C file that's parsed by doxygen. The file list does contain `vim.lua` though... that said, it's hard to show, since my browser does not pick up the css files for some reason and it all looks very bad. Maybe you should download the PR and run `doxygen src/Doxyfile && xdg-open buil/doxygen/html/index.html` and see what I mean, I'm lost with this.

> If lua2dox.lua is "generalized" we could include it in runtime/lua/ . What do you think?

Not sure, what would be the usecase for anyone? I also have doubts wrt it being robust enough for real user contact... let's see how things look like when it does what we want, and then think about that again.

KillTheMule commented:
I changed it to use `--@`. There's a feature about multiline doc comments, for now I've simply changed it to use `[[@` as well instead of `[[!`, but I'll need to test how this really works after I've figured out how to really produce the doxygen output shown on neovim.io.

KillTheMule commented:
Annoyingly, the troubles I had came from the fact that I was using doxygen 8.14 locally, but bot-ci uses 8.11. Changing the version fixes the layout, but makes the filter not work... I'll try to change stuff to work with 8.11.

KillTheMule commented:
Ok, got it, there's a bug somewhere that had me call doxygen 8.11 with `LANG=/usr/lib/locale/locale-archive`, which made the filter script fail, so I reset it with `LANG=""` in the filter script. Not sure if that workaround is needed anywhere else...

Other than that, [here](https://imgur.com/a/yaRVUY7) is how it looks like after annotating a function in [vim.lua](https://github.com/neovim/neovim/pull/9740/files#diff-c63607c08c91d28211f546c1937fa862R12). You can only see this stuff by going through the files list, which seems to be the case for all things C as well (other than the data structures). Should that be different? Should we try to make an entry in "Modules", something like "lua api" or so?  The remote API isn't accessible anyways, other than going through the files list to the correct files...

hkupty commented:
Maybe I'm looking from a very different perspective, but I've got the need to document my lua plugins and one of the best ways I found to do that was through [LDoc](https://stevedonovan.github.io/ldoc/manual/doc.md.html) and a [custom filter](https://github.com/clojure-vim/acid.nvim/blob/lua/wrapper/mdexport.lua) that writes to markdown.

An alternative to the doxygen issues would be to come up with a  similar filter that'd write to vim doc format. wdyt?

justinmk commented:
> Should we try to make an entry in "Modules", something like "lua api" or so?
> ...
> it's somewhat easy to create a module "Lua-API", but it puts an additional burden on the doc comments, because you have to use `--@ingroup Lua-API` so your function is mentioned there.

Good to know, but not necessary for now.

> Another disadvantage to the whole approach is that doxygen spews out errors about unknown parameters.

You mean, it flags `pid` even though you added `@param pid` to `_os_proc_info` ?

KillTheMule commented:
> You mean, it flags `pid` even though you added `@param pid` to `_os_proc_info` ?

Yes. There are several such errors for C code, they look legitimate (I'll send a PR to fix those some time), but we'll probably have to find a way to silence the errors about lua code.

KillTheMule commented:
So this generates entries in `api.txt` now. The [function parameters](https://github.com/neovim/neovim/commit/a7cc8859daf649e04f01cedcba93a5ff8df3b219#diff-49e08ed0add04457bc8a43c962736f13R1684) look off, I'll investigate.

Questions:
* I've added the lua file [manually](https://github.com/neovim/neovim/commit/a7cc8859daf649e04f01cedcba93a5ff8df3b219#diff-6c14a232993ad9487e0328d461e2a84cR82). How should we do that, generally?
* Not sure how [the section](https://github.com/neovim/neovim/commit/a7cc8859daf649e04f01cedcba93a5ff8df3b219#diff-49e08ed0add04457bc8a43c962736f13R1679) came into being. I'll found out, but what should it be?
* Do we want some sort of argument type documentation? I simply added a string to the comment [here](https://github.com/neovim/neovim/commit/a7cc8859daf649e04f01cedcba93a5ff8df3b219#diff-49e08ed0add04457bc8a43c962736f13R1684), but do we want something more formal and stringent?

KillTheMule commented:
Apart from actually writing the doc comments, this is where it should go, so I'm marking this RDY. I can write the comments in another PR or this one, I don't mind. Code-wise this feels pretty messy, but otoh it's not user-facing. The `gen_api_vimdoc.py` script mangles `api.txt` quite a bit, but my changes should not influence this... Should I squash this? I never know what's best, but from my POV this can just be one commit.

justinmk commented on scripts/gen_api_vimdoc.py:
```diff
@@ -277,9 +285,12 @@ def parse_source_xml(filename):
-                member.getAttribute('kind') != 'function':
+                member.getAttribute('kind') != 'function' or \
+                member.getAttribute('prot') == 'private' or \
+                get_text(get_child(member, 'name')).startswith('_'):
```
👍

justinmk commented on scripts/gen_api_vimdoc.py:
```diff
@@ -43,8 +43,10 @@
 # String used to find the start of the generated part of the doc.
-section_start_token = '*api-global*'
+api_section_start_token = '*api-global*'
+luadoc_section_start_token = '*api-lua*'
```
I think it's a good time to parameterize these config things.

```python
CONFIG = {
  'api': {
    'filename': 'api.txt',
    # String used to find the start of the generated part of the doc.
    'section_start_token': '*api-global*',
    'section_order' : [
      'vim.c',
      'buffer.c',
      'window.c',
      'tabpage.c',
      'ui.c',
    ],
  }
  'lua': {
    'filename': 'if_lua.txt'
    'section_start_token': '*api-lua*',
    'section_order' : [
      'vim.lua',
    ],
  }
```

justinmk commented on scripts/gen_api_vimdoc.py:
```diff
@@ -471,67 +492,100 @@ def gen_docs(config):
+        apidocs += section_doc
+        apidocs += '\n\n\n'
+
+    for filename in luadoc_section_order:
```
Rather than trying to do everything in one pass, can we avoid this extra/duplicate code by choosing the "mode" at invocation time? Then this will be something like:

    for filename in CONFIG[mode]["section_order"]:

justinmk commented on scripts/gen_api_vimdoc.py:
```diff
@@ -471,67 +492,100 @@ def gen_docs(config):
+            apidocs += '\n\n\n'
+
+    apidocs = apidocs.rstrip() + '\n\n'
+    apidocs += ' vim:tw=78:ts=8:ft=help:norl:\n'
```
Just pretend `apidocs` is "whatever docs we're currently working on". We don't want to have separate codepaths for every help doc we output.

justinmk commented on scripts/gen_api_vimdoc.py:
```diff
@@ -471,67 +492,100 @@ def gen_docs(config):
 INPUT                  = {input}
 INPUT_ENCODING         = UTF-8
-FILE_PATTERNS          = *.h *.c
+FILE_PATTERNS          = *.h *.c *.lua
```
Might want to set this based on the "mode". Thus:

    FILE_PATTERNS             = "{file_patterns}"
    ...
    
then later in `gen_docs()` ...

```
    p.communicate(config.format(input=src_dir, output=out_dir,
                                file_patterns=CONFIG[mode]["file_patterns"],
                                filter=filter_cmd).encode('utf8')) 
```

justinmk commented on src/nvim/lua/vim.lua:
```diff
@@ -163,9 +163,10 @@ local function trim(s)
 
-local deepcopy
+local function deepcopy(orig)
+end
```
maybe add `error()` as the function body.

justinmk commented:
>  I can write the comments in another PR or this one, I don't mind.

As an exercise for this PR, might as well move the existing docs from if_lua.txt into docstrings, to see how it looks. But other undocumented functions, can wait for another PR.

> Code-wise this feels pretty messy, but otoh it's not user-facing

The script will see further enhancements because we eventually will generate other docs. So I made some suggestions for maintainability.

> script mangles api.txt quite a bit

Don't worry, it's because api.txt was manually changed recently. I am working on a fix.

> Should I squash this? I never know what's best

Yes. For delicate/risky changes where we want to "save" the state of a review, squashing during review is risky. But this PR is not risky.

justinmk commented on src/nvim/lua/vim.lua:
```diff
@@ -118,6 +118,19 @@ local function _update_package_paths()
+---(passed as an argument to String.find)
+--@returns An iterator over the split components
+---
+--- Split a string by a given separator.
```
Can the main description go at the top? Or this is a a limitation of lua2dox ?

KillTheMule replied:
I don't think it's an issue of lua2dox, but rather I can't make doxygen show up stuff the way I want. Check out the latest 2 commits.

justinmk commented:
> Try to put examples after params, but keep short description before it

This isn't going to work, Doxygen has no way to know if the latter part should continue the `@returns` block or start a new element. It's better (and more common) to put the entire description at the top, before `@param`, `@return`, etc.

KillTheMule commented:
I've removed the code duplication, please have a look. What's left is: Fix the doc comments as you suggested, and remove the duplicate `vim.*` docs because we're autogenerating them now.

justinmk commented:
> I've removed the code duplication, please have a look.

LGTM. I think the `lua2dox_filter` bash script could be reduced to ~3 lines, but that's not important. 

Also later we should maybe lint/format the `lua2dox.lua` file, but for now just making it 2-space indent would be a big improvement.

> Question: Is [this](https://github.com/neovim/neovim/blob/master/runtime/doc/if_lua.txt#L390) still accurate?

No!  `vim.inspect` works even with `nvim -u NONE`. Cool :) So this line can be deleted:

> Those functions are only available if the runtime files have been loaded. In particular, they are not available when using `nvim -u NONE`.

KillTheMule committed:
gen_vimdoc.py: get Lua docs via lua2dox.lua #9740

KillTheMule committed:
Document the vim.lua functions

KillTheMule commented:
After obsessing a bit about the right squashing I think this is done now. I could not make URLs work prettily for some reason, not sure why doxygen deletes the whitespaces before them... anyways, docs look good imho, I shifted them around a bit too to accomdate the fact that we're adding the generated docs at the end.

justinmk committed:
gen_vimdoc.py: support lua/shared.lua module [ci skip]

justinmk committed:
lua/shared: share deepcopy() with test/*

deepcopy() was duplicated in test/helpers.lua

justinmk committed:
lua/shared: share more stuff

Leave trim() in vim.lua, because gen_vimdoc.py needs at least one
function in there, else it gets confused...

justinmk commented:
- Fixed duplicate tag, which causes CI to fail.
- Moved most functions to `shared.lua` (see https://github.com/neovim/neovim/pull/9301 )
- Various other changes.

h-michael commented on scripts/lua2dox.lua:
```diff
@@ -0,0 +1,666 @@
+          this:warning(inStream:getLineNo(),'something weird here')
+        end
+        fn_magic = nil -- mustn't indavertently use it again
+      elseif string.find(line,'=%s*class%(') then
```
@justinmk 
Do you remember what purpose this code was added to?
At present, there seems to be no code that passes this block in the Lua code in Neovim.

KillTheMule replied:
No, sorry, I don't (and didn't) have a good understanding of how lua2dox really worked. Just enough to hack in a bit...

h-michael replied:
No problem.:)  I will make a few modifications for #11276.
